rama-core/src/matcher/ext.rs

```diff
@@ -0,0 +1,78 @@
+use crate::{
+    matcher::{Extensions, Matcher},
+    Context,
+};
+use std::any::Any;
+
+pub struct ExtensionMatcher<T: Send + Sync> {
+    marker: T,
+}
+
+struct Const<T: Send + Sync>(T);
+struct FnBox(Box<dyn Fn(&dyn Any) -> bool + Send + Sync>);
+
+impl ExtensionMatcher<FnBox> {
+    pub fn with<T: Send + Sync + 'static, F: Fn(&T) -> bool + Send + Sync + 'static>(f: F) -> Self {
+        let wrapped_fn = move |v: &dyn Any| {
+            if let Some(concrete) = v.downcast_ref::<T>() {
+                f(concrete)
+            } else {
+                false
+            }
+        };
+        Self {
+            marker: FnBox(Box::new(wrapped_fn)),
+        }
+    }
+}
+
+impl<T: Send + Sync + Sync + std::cmp::PartialEq> ExtensionMatcher<Const<T>> {
+    pub fn constant(value: T) -> Self {
+        Self {
+            marker: Const(value),
+        }
+    }
+}
+
+impl<T: Send + Sync + std::cmp::PartialEq + 'static, State, Request> Matcher<State, Request>
```

Comment:
Thanks to the private trait trick you can now just implement the `ExtensionMatcher` with just 1 impl block:

```rust
impl<State, Request, P, T> Matcher<State, Request> for ExtensionMatcher<P, T>
where
    State: Clone + Send + Sync + 'static,
    Request: Send + 'static,
    T: Clone + Send + Sync + 'static,
    P: private::ExtensionPredicate<T>,
{
    fn matches(&self, _ext: Option<&mut Extensions>, ctx: &Context<State>, _req: &Request) -> bool {
        ctx.get::<T>()
            .map(|v| self.predicate.call(v))
            .unwrap_or_default()
    }
}
```